### ISSUE TITLE
Upgrading language client causing Error [ERR_STREAM_WRITE_AFTER_END]: write after end, message cannot be parsed

### ISSUE THREAD

**Issue by SumayyaZiyad** [microsoft/vscode-languageserver-node#755]
Upgraded the vscode-languageclient package from 5.1.1 to the latest and the following errors are now being thrown.

![Screenshot from 2021-03-27 19-23-35](https://user-images.githubusercontent.com/36619740/112722937-30839a80-8f32-11eb-8837-514e41c29223.png)

**Comment by rcjsuen** [comment]
> Upgraded the vscode-languageclient package from 5.1.1 to the latest and the following errors are now being thrown.

1. Is your language server and/or VS Code extension code available publicly somewhere?
2. What version did you upgrade `vscode-languageclient` from?

**Comment by SumayyaZiyad** [comment]
From 5.1.1. 
It fails at client.onReady().

![Screenshot from 2021-03-27 22-24-40](https://user-images.githubusercontent.com/36619740/112728044-494c7a00-8f4b-11eb-8e05-7efce7112a17.png)

Once the initalization is done, before running commands i check whether the client is ready, and that is where the error is thrown.
All of this was working perfectly prior to the package upgrade.

**Comment by rcjsuen** [comment]
> From 5.1.1.

Sorry, I see now I misread your original comment. So you jumped from 5.1.1 all the way to `7.1.0-next.4`? Did you try upgrading to 6.x first?

Is your code hosted at [ballerina-platform/plugin-vscode](https://github.com/ballerina-platform/plugin-vscode)?

**Comment by SumayyaZiyad** [comment]
> Sorry, I see now I misread your original comment. So you jumped from 5.1.1 all the way to `7.1.0-next.4`? Did you try upgrading to 6.x first?
To 7.0.0. Didn't try 6.x yet. Will try that also.


> Is your code hosted at [ballerina-platform/plugin-vscode](https://github.com/ballerina-platform/plugin-vscode)?
No this is a different project. But the server implementation is the same. So I am assuming that upgrading the langclient of that project also gives the same prob.

**Comment by SumayyaZiyad** [comment]
6.1.4 also works smoothy. Its just 7.x.x that breaks it. But I went through the documentation and noticed that for the 7.x.x updates it was just the classification of node/common/browser and few other changes that had happened IINM?

**Comment by danieldietrich** [comment]
I experience the same during LSP client/server connection initialization when using `7.0.0`. With `6.1.4` all works fine.

<img width="1409" alt="Screen Shot 2021-03-28 at 18 13 02" src="https://user-images.githubusercontent.com/743833/112804832-261de980-9075-11eb-8cd2-3da99041e78f.png">

Here is a part of the LSP payload that was sent to the server:

<img width="550" alt="Screen Shot 2021-03-28 at 21 16 17" src="https://user-images.githubusercontent.com/743833/112804856-2f0ebb00-9075-11eb-9251-72a20123a478.png">

If I patch the message (by setting `1` to `true`), there occurs another error:

<img width="1401" alt="Screen Shot 2021-03-28 at 22 01 23" src="https://user-images.githubusercontent.com/743833/112805337-c411b400-9075-11eb-9368-1ccd11255887.png">

**Comment by danieldietrich** [comment]
I recognized that some imports aren't found anymore when upgrading from 6.1.4 to 7.0.0.

**Should I create a new issue** or are these backward incompatible changes intended because of the major version jump from 6 to 7?

![Screen Shot 2021-03-29 at 11 15 09](https://user-images.githubusercontent.com/743833/112814921-122bb500-9080-11eb-8108-cce5116ee545.png)

**Comment by SumayyaZiyad** [comment]
There were changes made in 7.x.x that requires you to specify the code for each module depending on whether it is for node, browser or common.
In your case, you import should now be from "vscode-languageclient/node".
Checkout [this](https://github.com/Microsoft/vscode-languageserver-node/blob/main/README.md#3160-protocol-600-json-rpc-700-client-and-700-server) for details.

**Comment by dbaeumer** [comment]
@SumayyaZiyad can you provide any steps on how to reproduce this. Without being able to debug this it is very hard to tell why it is failing. And have you tried one of the official 7.0.x version and not the next version?

**Comment by SumayyaZiyad** [comment]
It is the official 7.0.0 version. Did NOT use the next version.
Basically what my code does is:
1) Initialize a client with a ID, name, clientOptions and serverOptions - with NO specification about Language features like completions and code actions
2) Start it
3) And before registering my extension commands, it checks whether the client is ready.

It is at this onReady that the code breaks, and the errors are thrown.

**Comment by dbaeumer** [comment]
I use comparable code here: https://github.com/microsoft/vscode-eslint/blob/main/client/src/extension.ts#L1919 and it works without problems. So I would really appreciate if you can provide me with detailed steps to reproduce (including the code) or even better a GitHub repository I can clone that demos what you are experiencing.

**Comment by dbaeumer** [comment]
I even added a dummy command registration to my readyHandler like this `Commands.registerCommand('helloWorld', () => {});` and everything still works. So there must be something you do different.

**Comment by SumayyaZiyad** [comment]
https://github.com/SumayyaZiyad/plugin-syntaxtree-visualizer/blob/2cea7febfd1d8aac224b975b3ed9c4b203e1b2c2/src/core/extension.ts#L167 this is where my issue originates when the client is updated to 7.x.

**Comment by dbaeumer** [comment]
@SumayyaZiyad that call doesn't look suspicious at all. As said earlier it is very hard to tell what is going wrong without steps on how to reproduce this.

**Comment by SumayyaZiyad** [comment]
Sorry I think I wasnt clear. That is the project for which I am trying to upgrade the client version. This when updated and run, throws the errors at the highlighted line.

**Comment by dbaeumer** [comment]
@SumayyaZiyad Thanks. I was hoping for something that is easier for me to execute. I don't have the tools installed to build the repository. Can you provide me with a repository having all the bits I need.

Have you tried to reproduce this using the simple lsp-sample here: https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-sample. That would make tracking this down a lot easier.

Alternatively I can give you instruction on how to debug this yourself.

**Comment by dbaeumer** [comment]
Here are the steps:

- compile the extension without web packing it. Will make debugging a lot easier.
- the actual writing happens here: https://github.com/microsoft/vscode-languageserver-node/blob/d58c00bbf8837b9fd0144924db5e7b1c543d839e/jsonrpc/src/common/messageWriter.ts#L147
- I recommend to set a breakpoint in the corresponding JS file inside the node_modules folder.

I suspect that for some reason the server is not available anymore when you try to write into the connection hence the write fails. You can check if the server is still running using a process explorer.

Let me know how it goes.

**Comment by SumayyaZiyad** [comment]
It seems like the `Error [ERR_STREAM_WRITE_AFTER_END]: write after end` is what is breaking the flow.
Attaching the stack trace for reference.

> Error [ERR_STREAM_WRITE_AFTER_END]: write after end
>     at writeAfterEnd (_stream_writable.js:266:14)
>     at Socket.Writable.write (_stream_writable.js:315:5)
>     at /home/sumayya/Documents/Project/plugin-syntaxtree-visualizer/node_modules/vscode-jsonrpc/lib/node/ril.js:90:29
>     at new Promise (<anonymous>)
>     at WritableStreamWrapper.write (/home/sumayya/Documents/Project/plugin-syntaxtree-visualizer/node_modules/vscode-jsonrpc/lib/node/ril.js:80:16)
>     at StreamMessageWriter.doWrite (/home/sumayya/Documents/Project/plugin-syntaxtree-visualizer/node_modules/vscode-jsonrpc/lib/common/messageWriter.js:100:33)
>     at /home/sumayya/Documents/Project/plugin-syntaxtree-visualizer/node_modules/vscode-jsonrpc/lib/common/messageWriter.js:91:29
>

**Comment by dbaeumer** [comment]
@SumayyaZiyad the interesting question is why the write fails. Is the server still running. Have you tried to set a breakpoint here https://github.com/microsoft/vscode-languageserver-node/blob/dbaeumer/workspacePull2/jsonrpc/src/node/ril.ts#L114 to find out if end is called. 

Checking if the stream is closed otherwise by having a breaking in the listeners might be helpful as well:

https://github.com/microsoft/vscode-languageserver-node/blob/dbaeumer/workspacePull2/jsonrpc/src/common/messageWriter.ts#L73

**Comment by andreasmuelder** [comment]
same here when updating from 6.1.1 to 7.0.0

**Comment by dbaeumer** [comment]
@andreasmuelder any steps on how to reproduce this best via a GitHub repository I can clone.
